src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -496,9 +497,8 @@ public Object getIndex(VirtualFrame frame, DynamicObject string, int index, NotP
         }
 
         @Specialization(guards = { "!isRubyRange(index)", "!isRubyRegexp(index)", "!isRubyString(index)" })
-        public Object getIndex(VirtualFrame frame, DynamicObject string, Object index, NotProvided length,
-                @Cached("new()") SnippetNode snippetNode) {
-            return getIndex(frame, string, (int) snippetNode.execute(frame, "Truffle::Type.rb_num2int(v)", "v", index), length);
+        public Object getIndex(VirtualFrame frame, DynamicObject string, Object index, NotProvided length) {
+            return getIndex(frame, string, (int) getSnippetNode().execute(frame, "Truffle::Type.rb_num2int(v)", "v", index), length);
```

Comment:
I think there is a hidden requirement here that the SnippetNode code must be the same Java String:
https://github.com/graalvm/truffleruby/blob/b619adec2f8056afc34e90fb852ed1a1c155eeac/src/main/java/org/truffleruby/language/SnippetNode.java#L100
But it probably works out because they are the same strings in this node and literal Java Strings are interned.

I think it would be clearer to extract the `getSnippetNode().execute` in its own method (`toInt` maybe?), and then it would be clear it's always the same code passed to it.

Reply:
It's also related to how we use the lazy pattern (always include the `execute` call in the method), see https://github.com/graalvm/truffleruby/blob/master/doc/contributor/code-patterns.md#the-lazy-pattern for details.

Reply:
@eregon I don't understand what 'which includes the `execute*` call on the helper node' means. What does 'includes' mean and why is it emphasised? How would you not include a call to `execute`?

Perhaps `SnippetNode` should specialise on the interned string being called? I don't see why it isn't doing that already. Can anyone remember any reasons?

Reply:
I think that means to add the execute method into the getter method; the pattern should be `toStr(value)`, not `getToStrNode().execute(value)`

Reply:
@chrisseaton Is it not clear in the documentation? Basically `getSnippetNode` => `toInt` (so the helper method for the lazy pattern also calls the node and doesn't just lazily instantiate it).

We'd need an extra inline cache there and make the implementation quite more complex.
The pattern was always to use one SnippetNode per snippet.
Anyway, I want to remove most SnippetNode usages as they should just be calls, but this is out of scope of this PR.